### PR TITLE
Remove unused headers in core

### PR DESCRIPTION
**Pull request by YYF233333**
Trying to apply [include-what-you-use](https://github.com/include-what-you-use/include-what-you-use) to godot codebase, but turns out it is not smart enough for such complex dependency graph, so I pick some safe changes and applied manually.

Hopefully this will reduce build time a bit. (and also warnings in your IDE) 😄

### CODE REVIEW

**Comment by akien-mga** [comment]
Seems like https://github.com/godotengine/godot/actions/runs/12396780139/job/34605644487?pr=100564 requires `script_language.h` in `main.cpp`.

**Review by akien-mga — APPROVED** [review]

**Comment by Repiteo** [comment]
Thanks!

File: core/os/thread.h [hunk 1]
```diff
@@ -42,8 +42,6 @@
 #include "core/templates/safe_refcount.h"
 #include "core/typedefs.h"
 
-#include <new>
```
Comment (akien-mga) [inline core/os/thread.h:45]: FYI, this was actually needed and caused a crash with GCC 12: #100768 (fixed in #101004).
This was a honest mistake but shows we need to be somewhat cautious when reviewing IWYU suggestions.

File: core/os/thread_safe.h [hunk 2]
```diff
@@ -31,8 +31,6 @@
 #ifndef THREAD_SAFE_H
 #define THREAD_SAFE_H
 
-#include "core/os/mutex.h"
```
Comment (akien-mga) [inline core/os/thread_safe.h:34]: This was also a wrong suggestion from IWYU, as it's used just below in macros. Apparently it doesn't support macros, so we need to be cautious there too when reviewing changes.

**Pull request merged**